Add logging util lib for Proto Message Logging (PML) http filter

dchakarwarti opened the pull request:
Commit Message: Add logging util lib for Proto Message Logging (PML) http filter
Additional Description: This functionality will be used in the PML filter
Risk Level: Low
Testing: Added unit tests
Docs Changes: The proto for PML filter is present at api/envoy/extensions/filters/http/proto_message_logging/v3/config.proto
Release Notes: Nil
Platform Specific Features: Nil

dchakarwarti commented:
@moderation

moderation commented on bazel/repository_locations.bzl:
```diff
@@ -987,6 +987,51 @@ REPOSITORY_LOCATIONS_SPEC = dict(
+        license = "Apache-2.0",
+        license_url = "https://github.com/google/or-tools/blob/stable/LICENSE",
+    ),
+    com_google_proto_field_extraction = dict(
+        project_name = "proto-field-extraction",
+        project_desc = "Library that supports the extraction from protobuf binary",
+        project_url = "https://github.com/grpc-ecosystem/proto-field-extraction",
+        version = "598807946f248917db4b7d2e7249de3e4eee0003",
+        sha256 = "791501e8ba94b91017ecba56e17e83d8de828ba883ed559ba7cc47229ca3b31f",
+        strip_prefix = "proto-field-extraction-{version}",
+        urls = ["https://github.com/grpc-ecosystem/proto-field-extraction/archive/{version}.zip"],
+        use_category = ["dataplane_ext"],
+        extensions = ["envoy.filters.http.grpc_json_transcoder", "envoy.filters.http.grpc_field_extraction"],
+        release_date = "2024-03-22",
+        cpe = "N/A",
+        license = "Apache-2.0",
+        license_url = "https://github.com/grpc-ecosystem/proto-field-extraction/blob/{version}/LICENSE",
+    ),
```
```
scorecard --repo github.com/grpc-ecosystem/proto-field-extraction
RESULTS
-------
Aggregate score: 3.1 / 10

Check scores:
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
|  SCORE  |          NAME          |             REASON             |                                               DOCUMENTATION/REMEDIATION                                               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts       | no binaries found in the repo  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#binary-artifacts       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Branch-Protection      | branch protection not enabled  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#branch-protection      |
|         |                        | on development/release         |                                                                                                                       |
|         |                        | branches                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | CI-Tests               | 0 out of 6 merged PRs          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#ci-tests               |
|         |                        | checked by a CI test -- score  |                                                                                                                       |
|         |                        | normalized to 0                |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | CII-Best-Practices     | no effort to earn an OpenSSF   | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#cii-best-practices     |
|         |                        | best practices badge detected  |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 6 / 10  | Code-Review            | found 3 unreviewed changesets  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#code-review            |
|         |                        | out of 9 -- score normalized   |                                                                                                                       |
|         |                        | to 6                           |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Contributors           | 0 different organizations      | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#contributors           |
|         |                        | found -- score normalized to 0 |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Dangerous-Workflow     | no workflows found             | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#dangerous-workflow     |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Dependency-Update-Tool | no update tool detected        | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#dependency-update-tool |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Fuzzing                | project is not fuzzed          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#fuzzing                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | License                | license file detected          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#license                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Maintained             | 0 commit(s) out of 11 and 0    | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#maintained             |
|         |                        | issue activity out of 0 found  |                                                                                                                       |
|         |                        | in the last 90 days -- score   |                                                                                                                       |
|         |                        | normalized to 0                |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Packaging              | no published package detected  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#packaging              |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Pinned-Dependencies    | no dependencies found          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#pinned-dependencies    |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | SAST                   | SAST tool is not run on all    | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#sast                   |
|         |                        | commits -- score normalized to |                                                                                                                       |
|         |                        | 0                              |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Security-Policy        | security policy file not       | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#security-policy        |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Signed-Releases        | no releases found              | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#signed-releases        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Token-Permissions      | no github tokens found         | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#token-permissions      |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Vulnerabilities        | no vulnerabilities detected    | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#vulnerabilities        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
```

moderation commented on bazel/repository_locations.bzl:
```diff
@@ -987,6 +987,51 @@ REPOSITORY_LOCATIONS_SPEC = dict(
+        license = "Apache-2.0",
+        license_url = "https://github.com/grpc-ecosystem/proto-field-extraction/blob/{version}/LICENSE",
+    ),
+    com_google_protoprocessinglib = dict(
+        project_name = "proto_processing_lib",
+        project_desc = "Library that provides utility functionality for proto field scrubbing",
+        project_url = "https://github.com/grpc-ecosystem/proto_processing_lib",
+        version = "09e6cf5da15c8b9d32c6706b2517890200110bc8",
+        sha256 = "38c0a012cbf6ab98615c398f775f07b19addd03442c2811ea35a878f671aedbc",
+        strip_prefix = "proto_processing_lib-{version}",
+        urls = ["https://github.com/grpc-ecosystem/proto_processing_lib/archive/{version}.zip"],
+        use_category = ["dataplane_ext"],
+        extensions = ["envoy.filters.http.grpc_json_transcoder", "envoy.filters.http.grpc_field_extraction"],
+        release_date = "2024-06-21",
+        cpe = "N/A",
+        license = "Apache-2.0",
+        license_url = "https://github.com/grpc-ecosystem/proto_processing_lib/blob/{version}/LICENSE",
+    ),
```
```
scorecard --repo github.com/grpc-ecosystem/proto_processing_lib
RESULTS
-------
Aggregate score: 3.4 / 10

Check scores:
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
|  SCORE  |          NAME          |             REASON             |                                               DOCUMENTATION/REMEDIATION                                               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts       | no binaries found in the repo  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#binary-artifacts       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Branch-Protection      | branch protection not enabled  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#branch-protection      |
|         |                        | on development/release         |                                                                                                                       |
|         |                        | branches                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | CI-Tests               | no pull request found          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#ci-tests               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | CII-Best-Practices     | no effort to earn an OpenSSF   | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#cii-best-practices     |
|         |                        | best practices badge detected  |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 1 / 10  | Code-Review            | found 9 unreviewed changesets  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#code-review            |
|         |                        | out of 11 -- score normalized  |                                                                                                                       |
|         |                        | to 1                           |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 3 / 10  | Contributors           | 1 different organizations      | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#contributors           |
|         |                        | found -- score normalized to 3 |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Dangerous-Workflow     | no workflows found             | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#dangerous-workflow     |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Dependency-Update-Tool | no update tool detected        | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#dependency-update-tool |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Fuzzing                | project is not fuzzed          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#fuzzing                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | License                | license file detected          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#license                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 5 / 10  | Maintained             | 7 commit(s) out of 11 and 0    | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#maintained             |
|         |                        | issue activity out of 0 found  |                                                                                                                       |
|         |                        | in the last 90 days -- score   |                                                                                                                       |
|         |                        | normalized to 5                |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Packaging              | no published package detected  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#packaging              |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Pinned-Dependencies    | no dependencies found          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#pinned-dependencies    |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | SAST                   | no SAST tool detected          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#sast                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Security-Policy        | security policy file not       | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#security-policy        |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Signed-Releases        | no releases found              | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#signed-releases        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Token-Permissions      | no github tokens found         | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#token-permissions      |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Vulnerabilities        | no vulnerabilities detected    | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#vulnerabilities        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
```

moderation commented on bazel/repository_locations.bzl:
```diff
@@ -987,6 +987,51 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         license = "Apache-2.0",
         license_url = "https://github.com/grpc-ecosystem/proto-field-extraction/blob/{version}/LICENSE",
     ),
+    com_google_ortools = dict(
+        project_name = "or-tools",
+        project_desc = "Software suite for combinatorial optimization.",
+        project_url = "https://github.com/google/or-tools",
+        version = "2384738a951561bf905f3435651310841c309b4c",
+        sha256 = "b374f96f4509ab8c6c185bb497f4284c3b309bcaf87d1478eb4ddda009d1f12d",
+        strip_prefix = "or-tools-{version}",
+        urls = ["https://github.com/google/or-tools/archive/{version}.zip"],
+        use_category = ["dataplane_ext"],
+        extensions = ["envoy.filters.http.grpc_json_transcoder", "envoy.filters.http.grpc_field_extraction"],
+        release_date = "2024-06-10",
+        cpe = "N/A",
+        license = "Apache-2.0",
+        license_url = "https://github.com/google/or-tools/blob/stable/LICENSE",
+    ),
```
Scorecard score - [4.8](https://scorecard.dev/viewer/?uri=github.com/google/or-tools)

kyessenov replied:
What's the purpose of combinatorial optimization for the extension? I'd be suspicious of using any algorithms here since they are typically super-linear.

kyessenov replied:
/wait-any

or-tools is research software, can you explain what it's used for?

dchakarwarti replied:
Hi Kuat!

`or-tools` isn't directly used in the logging libraries added for the PML filter. However, it is required for the Proto Processing Libraries. [Reference](https://github.com/grpc-ecosystem/proto_processing_lib/blob/09e6cf5da15c8b9d32c6706b2517890200110bc8/proto_processing_lib/proto_scrubber/BUILD.bazel#L94C11-L94C29).

The library is using `ortools/base/map_util.h`.

In envoy, the build fails if I do not explicitly import the `or-tools` library. 
```
$ bazel build //source/extensions/filters/http/proto_message_logging/logging_util:logging_util
ERROR: /usr/local/google/home/dchakarwarti/.cache/bazel/_bazel_dchakarwarti/0a1ddbe361fc4585fd11ce5863817141/external/com_google_protoprocessinglib/proto_processing_lib/proto_scrubber/BUILD.bazel:73:11: no such package '@com_google_ortools//ortools/base': The repository '@com_google_ortools' could not be resolved: Repository '@com_google_ortools' is not defined and referenced by '@com_google_protoprocessinglib//proto_processing_lib/proto_scrubber:field_mask_node'
ERROR: Analysis of target '//source/extensions/filters/http/proto_message_logging/logging_util:logging_util' failed; build aborted: 
INFO: Elapsed time: 0.865s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (25 packages loaded, 840 targets configured)
```

kyessenov replied:
I think it was added because it's coming from Google internal gtl, and or-tools exported some fork of it at some point?  Can you re-work your dependency not to depend on a repo to use gtl helper - you can simply copy paste those two helpers directly into your code.

kyessenov replied:
A side question for these three proto libraries - are you somehow hooked into the internals of protobuf? Is there some hidden friend access? We can't update protobufs nearly as fast in OSS, so it'd be challenging if you use private proto interfaces.

dchakarwarti replied:
>  you can simply copy paste those two helpers directly into your code.

Sure, I removed the `ortools` dependency from here.

Additionally, I also removed `@com_google_proto_field_extraction` external repo dependency as it was already being used as `@com_google_protofieldextraction`.

Now, the only newly introduced dependency is of `@com_google_protoprocessinglib`.

>  are you somehow hooked into the internals of protobuf?

Actually, no. There is no hidden friend access. 

To add more context, this library is part of [grpc-ecosystem](https://github.com/grpc-ecosystem/proto_processing_lib) repo from which we have been using the field extraction and proto converter libraries as well.

nezdolik commented:
@dchakarwarti have you followed the [extension policy](https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md#adding-new-extensions)? Does this extension have a maintainer sponsor?

dchakarwarti commented:
> @dchakarwarti have you followed the [extension policy](https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md#adding-new-extensions)? Does this extension have a maintainer sponsor?

Yes, we have followed the extension policy. The config proto for the PML filter has already been submitted previously: https://github.com/envoyproxy/envoy/pull/31735.


@yanavlasov is our sponsor.

soulxu commented:
/assign @yanavlasov

soulxu commented:
@dchakarwarti it seems the code still have format issue.

dchakarwarti commented:
> Details

Thanks Alex! Handled the formatting related issues.

dchakarwarti commented:
cc: @yanavlasov

yanavlasov commented:
LGTM from me. Please merge main to fix the coverage build.

I will wait for @kyessenov approval and then merge.

dchakarwarti commented:
> LGTM from me. Please merge main to fix the coverage build.
> 
> I will wait for @kyessenov approval and then merge.

Sure, addressed the coverage build.

Added some more unit tests. Requesting a review please.

Thank you!

dchakarwarti commented:
@kyessenov can you please take a look at the dependency changes?

kyessenov reviewed (approved):
Approving deps change - don't have the full context for this filter. Thanks for addressing the extraneous import.
